File: APIs/python/InstanceAPI.py
import socket

# This api let's you use the api as an instance of voxel display, not as a class you have to extend
# This means that YOU have to supply your own loops and threads
class VoxelDisplay():
  def __init__ ( self, debug=False ):
    self.voxels = set()
    self.debug = debug

    self.sock = socket.socket( socket.AF_INET, socket.SOCK_STREAM)
    self.sock.connect( ("localhost", 8999 ) )

  def flush( self ):
      line = ", ".join([ " ".join( str(i) for i in v) for v in self.voxels ])

      if (self.debug):
        print(line)
      self.sock.send( bytes(line  + "\n", 'UTF-8') )

  def setVoxel( self, x, y, z ):
    self.voxels.add( (x,y,z) )

  def toggleVoxel( self, x, y, z ):
    self.voxels ^= set( [(x,y,z)] )

  def getState( self ):
    return self.voxels

  def clearState( self ):
    self.voxels = set()

File: APIs/ruby/InstanceAPI.rb
require 'socket'
require 'set'
include Socket::Constants

class VoxelDisplay

  def initialize(debug=false)
    @voxelSet = Set.new()
    @debug = debug

    @socket = Socket.new( AF_INET, SOCK_STREAM, 0 )
    sockaddr = Socket::pack_sockaddr_in( 8999, '127.0.0.1' )
    @socket.connect( sockaddr )
    puts "CONNECTED"
  end
  
  def set_voxel( x, y, z )
    @voxelSet.add [x, y, z]
  end

  def toggle_voxel( x, y, z )
    @voxelSet ^= Set.new([[x,y,z]])
  end

  def clear_state(normal=true)
    @voxelSet = Set.new
  end

  def get_state
    @voxelSet
  end

  def flush
    res = @voxelSet.map {|e| e.join(" ")} ::join(", ")
    if @debug
      puts res
    end
    @socket.write(res+"\n")
  end

end
